Plutarch/Internal/PlutusType.hs

```diff
@@ -1,317 +1,125 @@
-    ( code ~ Code (a s)
-    , pcode ~ PCode s a
-    , PGeneric s a
-    , AppL b (ScottList' s pcode b)
-    , GPMatch a 0 code b s
-    , PInner a b ~ ScottFn (ScottList' s pcode b) b
-    ) =>
-    (Term s (PInner a b)) ->
-    (a s -> Term s b) ->
-    Term s b
-  pmatch' x f = gpmatch @a x (f . gpto)
-
-instance {-# OVERLAPPABLE #-} PlutusType a => PMatch a where
-  pmatch x f = pmatch' (punsafeCoerce x) f
-
-instance PlutusType a => PCon a where
-  pcon x = punsafeCoerce (pcon' x)
-
-class PCon a where
-  -- | Construct a Plutarch Term via a Haskell datatype
-  pcon :: a s -> Term s a
-
-class PMatch a where
-  -- | Pattern match over Plutarch Terms via a Haskell datatype
-  pmatch :: Term s a -> (a s -> Term s b) -> Term s b
-
--- | Generic version of `pcon'`
-gpcon ::
-  forall a c s code pcode.
-  ( PlutusType a
-  , PGeneric s a
-  , code ~ Code (a s)
-  , pcode ~ PCode s a
-  , GPCon pcode c s
-  , PLamL (ScottList' s pcode c) c s
-  , ScottFn (ScottList' s pcode c) c ~ ScottFn' (ScottList s pcode c) c
-  ) =>
-  SOP (Term s) pcode ->
-  Term s (ScottFn' (ScottList s pcode c) c)
-gpcon (SOP val) =
-  plamL @(ScottList' s pcode c) @c $ \(f :: NP (Term s) (ScottList' s pcode c)) ->
-    gpcon' @pcode @c @s f val
-
-{- |
-  `gpcon'`, given a *partial* scott encoding (as a `PLamL`) and a sum choice, applies
-  that encoding to the sum choice.
-
-  The partial encoding is any tail of the full scott encoded function, such that
-  one of its elements corresponds to the sum choice.
--}
-class GPCon (xss :: [[PType]]) (c :: PType) (s :: S) where
-  gpcon' :: NP (Term s) (ScottList' s xss c) -> NS (NP (Term s)) xss -> Term s c
-
-instance GPCon '[] c s where
-  gpcon' Nil = \case {}
-
-instance (GPCon xs c s, AppL c x) => GPCon (x ': xs) c s where
-  gpcon' (f :* fs) = \case
-    Z x -> f `appL` x
-    S xs -> gpcon' fs xs
-
-{- |
-  Generic version of `pmatch'`
--}
-gpmatch ::
-  forall a s c code pcode.
-  ( PGeneric s a
-  , code ~ Code (a s)
-  , pcode ~ PCode s a
-  , AppL c (ScottList' s pcode c)
-  , GPMatch a 0 code c s
-  ) =>
-  Term s (ScottFn (ScottList' s pcode c) c) ->
-  (SOP (Term s) pcode -> Term s c) ->
-  Term s c
-gpmatch x f =
-  x `appL` gpmatch' @a @0 @code @c @s f
-
-{- |
-  `gpmatch'` returns a hlist of lambdas (or delayed terms) to be applied on the
-  scott encoding function.
--}
-class GPMatch (a :: PType) (n :: Nat) (xss :: [[Type]]) (c :: PType) (s :: S) where
-  gpmatch' :: (SOP (Term s) (PCode s a) -> Term s c) -> NP (Term s) (ScottList' s (ToPType2 xss) c)
-
-instance GPMatch a n '[] c s where
-  gpmatch' _ = Nil
-
-instance
-  ( code ~ Code (a s)
-  , ToPType xs ~ IndexList n (PCode s a)
-  , GPMatch a (n + 1) xss c s
-  , PLamL (ToPType xs) c s
-  , MkNS n (PCode s a) (NP (Term s))
-  , All Top (ToPType xs)
-  , All Top xs
+import Plutarch.Internal.Quantification (PFix (PFix), PForall (PForall), PSome (PSome))
+import Plutarch.Internal.Witness (witness)
+
+class PlutusTypeStrat (strategy :: Type) where
```

Comment:
what was the reason behind choosing making the `strategy` open kinded? is there something potentially bad with using data kinds or are you just leaving room for expansion by users?

Reply:
There's no reason to make it closed, and it would make it more complex.